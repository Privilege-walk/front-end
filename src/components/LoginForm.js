import React, { useState } from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import axios from 'axios';
import { Navigate } from "react-router-dom";
import "./App.css";

export default function LoginForm() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [login, setLogin] = useState(false);
  const [errMsg, setErrMsg] = useState("");

  function validateForm() {
    return username.length > 0 && password.length > 0;
  }

  function handleSubmit(event) {
    event.preventDefault();
    axios.post(`http://localhost:8000/auth/login/`, { username, password })
      .then(res => {
        console.log(res);
        console.log(res.data);
        if (res.data.status === true) {
            setLogin(true);
        } else {
            setErrMsg("Please check your username or password!");
        }
      })
  }

  return (
    <div className="Login">
      <Form onSubmit={handleSubmit}>
        <Form.Group size="lg" controlId="username">
          <Form.Label>Username</Form.Label>
          <Form.Control
            autoFocus
            type="username"
            value={username}
            onChange={(e) => setUsername(e.target.value)}
          />
        </Form.Group>
        <Form.Group size="lg" controlId="password">
          <Form.Label>Password</Form.Label>
          <Form.Control
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </Form.Group>
        <Button size="lg" type="submit" 
          disabled={!validateForm()}
        >
          Login
        </Button>
        <div className="errMsg">{errMsg}</div>
      </Form>
      {
          login &&
          <Navigate to="/loggedin" />
      }
    </div>
  );
}