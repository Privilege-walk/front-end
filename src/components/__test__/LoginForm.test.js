import React from "react";
import { render, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';
import MockAdapter from "axios-mock-adapter";
import LoginForm from '../LoginForm';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk'
// import { loginUser } from "../../Store/actions";
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    Navigate: () => "Navigate",
    useNavigate: () => mockedUsedNavigate,
}))

const mockLoginUser = jest.fn();
jest.mock('../../Store/actions', () => ({
    loginUser: () => mockLoginUser
}));

const setup = () => {
    let utils;
    const store = mockStore({
        loginUser: mockLoginUser,
    });
    act (() => {
        utils = render(
        <Provider store={store}>
            <LoginForm />
        </Provider>
        );
    }) 
    const usernameInput = screen.getByLabelText('Username');
    const passwordInput = screen.getByLabelText('Password');
    const submitBtn = screen.getByText('Login', {selector: 'button'});
    const errDiv = screen.getByTestId('loginErrorId');
    return {
        usernameInput,
        passwordInput,
        submitBtn,
        errDiv,
        ...utils
    }
}
test('renders without crashing',  ()=> {
    const utils = setup(); 
    const { asFragment, usernameInput, passwordInput } = utils;
    expect(asFragment()).toMatchSnapshot();
    expect(usernameInput.value).toBe('');
    expect(passwordInput.value).toBe('');
});

test('Successful login',  async ()=> {
    const utils = setup(); 
    const { usernameInput, passwordInput, submitBtn, errDiv } = utils;
    const username = "user";
    const password = "password"
    var mock = new MockAdapter(axios);
    mock.onPost(
        '/auth/login/', 
        {username,  password}
    ).reply(200, {status: true, token: "token"});
    
    expect(submitBtn).toHaveProperty('disabled', true);
    fireEvent.change(usernameInput, {target: {value: username}});
    expect(submitBtn).toHaveProperty('disabled', true);
    fireEvent.change(passwordInput, {target: {value: password}});
    expect(submitBtn).toHaveProperty('disabled', false);
    
    mockLoginUser.mockReturnValueOnce({payload: {token: "token"}})
    expect(errDiv).toHaveTextContent("");
    await act(async () => {   
        await fireEvent.click(submitBtn);
    });
    expect(errDiv).toHaveTextContent("");
    
});

test('Failed login', async () => {
    const utils = setup(); 
    const { usernameInput, passwordInput, submitBtn, errDiv } = utils;
    const username = "user";
    const password = "password";
    
    expect(submitBtn).toHaveProperty('disabled', true);
    fireEvent.change(usernameInput, {target: {value: username}});
    expect(submitBtn).toHaveProperty('disabled', true);
    fireEvent.change(passwordInput, {target: {value: password}});
    expect(submitBtn).toHaveProperty('disabled', false);
    
    mockLoginUser.mockReturnValueOnce({payload: {token: ""}})
    expect(errDiv).toHaveTextContent("");
    await act(async () => {
        await fireEvent.click(submitBtn);
    });
    const errMessage = "Please check your username or password!";
    expect(errDiv).toHaveTextContent(errMessage);
    
});