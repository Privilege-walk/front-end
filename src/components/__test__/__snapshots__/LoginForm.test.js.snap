// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="Login"
  >
    <form
      class=""
    >
      <div
        class="mb-3"
        id="username"
      >
        <label
          class="form-label"
          for="username-input"
        >
          Username
        </label>
        <input
          class="form-control"
          id="username-input"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
        id="password"
      >
        <label
          class="form-label"
          for="password-input"
        >
          Password
        </label>
        <input
          class="form-control"
          id="password-input"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-secondary btn-lg disabled"
        disabled=""
        type="submit"
      >
        Login
      </button>
      <div
        class="errMsg"
        data-testid="loginErrorId"
      />
    </form>
  </div>
</DocumentFragment>
`;
